ksql-common/src/test/java/io/confluent/ksql/util/timestamp/StringToTimestampParserTest.java

```diff
@@ -0,0 +1,197 @@
+package io.confluent.ksql.util.timestamp;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import org.junit.Test;
+
+public class StringToTimestampParserTest {
+
+  @Test
+  public void shouldParseFullTimestamp() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd:HH");
+
+    // When:
+    final long ts = parser.parse("1605/11/05:12");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 5, 12, 0))));
+  }
+
+  @Test
+  public void shouldParseFullTimestampWithMinutes() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd HH:mm");
+
+    // When:
+    final long ts = parser.parse("1605/11/05 12:10");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 5, 12, 10))));
+  }
+
+  @Test
+  public void shouldParseWithOptionalElements() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd HH[:mm]");
+
+    // When:
+    final long ts = parser.parse("1605/11/05 12");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 5, 12, 0))));
+  }
+
+  @Test
+  public void shouldParseWithOptionalDefaultedFields() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM[/dd]");
+
+    // When:
+    final long ts = parser.parse("1605/11/05");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 5, 0, 0))));
+  }
+
+  @Test
+  public void shouldParseClockHourOfDay() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd[ k]");
+
+    // When:
+    final long ts = parser.parse("1605/11/01 2");
+
+    // Then:
+    assertThat(ts, equalTo(toMillis(LocalDateTime.of(1605, 11, 1, 2, 0))));
+  }
+
+  @Test
+  public void shouldParseWithTimeZone() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("yyyy/MM/dd HH z");
+
+    // When:
+    final long ts = parser.parse("1605/11/01 12 UTC");
+
+    // Then:
+    assertThat(ts, equalTo(
+        toMillis(ZonedDateTime.of(1605, 11, 1, 12, 0, 0, 0, ZoneId.of("UTC"))
+            .withZoneSameInstant(ZoneId.systemDefault())
+            .toLocalDateTime())));
+  }
+
+  @Test
+  public void shouldResolveDefaultsWithNoFormat() {
+    // Given:
+    final StringToTimestampParser parser = new StringToTimestampParser("");
```

Comment:
feels like this should be an error...

Reply:
Hehe it wasn't an error in the legacy code - what do you prefer, defaulting to something like `yyyy-mm-dd HH:MM` or throwing an error?

Reply:
What does the SQL syntax allow?  We should probably making it a parsing error first and foremost.  Then just have `StringToTimestampParser` throw on a empty format string.

However, given you ain't changing any functionality here feel free to leave this for another day if you're want to just get this merged.

Reply:
It seems that SQL standard has some default initialization parameter: https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions193.htm

I think in general this UDF is kinda out-of-line. It shouldn't have a `timeZone` parameter at all, and we should have a different UDF to convert time zones since that timeZone is ignored if the format has a timezone to begin with... I think I'll come up with a follow-up PR after this one to make this align a little bit more with SQL standard.